File: 216_Combination_Sum_III/216.py
class Solution:
    def combinationSum3(self, k, n):
        """
        :type k: int
        :type n: int
        :rtype: List[List[int]]
        """
        self.answer = []

        def func(path, remain, num):
            if len(path) == k:
                if num == 0:
                    self.answer.append(path)
            else:
                for i in range(len(remain) - k + len(path) + 1):
                    if num < remain[i]:
                        break
                    func(path + [remain[i]], remain[i + 1:], num - remain[i])
        func([], range(1, min(9, n) + 1), n)
        return self.answer

File: 216_Combination_Sum_III/216.cpp
class Solution {
public:
    typedef vector<vector<int>> ListOfList;
    void recursion(vector<int> &path, vector<int> &remain, int k, int n, ListOfList &ans){
        if(k == 0){
            if(n == 0)
                ans.push_back(path);
        }else{
            for(int i = 0; i < remain.size() && remain[i] <= n; ++i){
                auto newPath = path;
                newPath.push_back(remain[i]);
                vector<int> newRemain(remain.begin()+i+1, remain.end());
                recursion(newPath, newRemain, k - 1, n - remain[i], ans);
            }
        }
    }
    ListOfList combinationSum3(int k, int n) {
        ListOfList ans;
        vector<int> path;
        vector<int> remain;
        for(int i = 1; i < 10 && i < n; ++i)
            remain.push_back(i);
        recursion(path, remain, k, n, ans);
        return ans;
    }
};